robot/motion.c

```diff
@@ -0,0 +1,245 @@
+
+#define SIGN(x) (x > 0 ? 1 : -1)
+
+typedef struct {
+	float x;
+	float y;
+} point_t;
+
+struct interpoints {
+	point_t tan1;
+	point_t tan2;
+	float alpha1;
+	float alpha2;
+};
+
+static float distance(point_t a, point_t b) {
+	return sqrt(pow((b.x - a.x), 2) + pow((b.y - a.y), 2));
+}
+
+static point_t dir2vec(float dir) {
+	point_t result = {mcos(dir), msin(dir)};
+	return result;
+}
+
+static float scal(point_t a, point_t b) {
+	return a.x*b.x + a.y*b.y;
+}
+
+static point_t subs(point_t from, point_t to) {
+	point_t result;
+	result.x = to.x - from.x;
+	result.y = to.y - from.y;
+	return result;
+}
+
+static float det(point_t a, point_t b) {
+	return a.x*b.y - b.x*a.y;
+}
+
+static void computeTangent(point_t *center1, int r1, point_t *center2, int r2, int i, point_t *tan1, point_t *tan2) {
+	point_t h;
+	int l1 = i % 2 ? 1 : -1;
+	int l2 = i > 1 ? 1 : -1;
+	int innerTan = -l1*l2;
+
+	h.x = (r1*center2->x + innerTan*r2*center1->x) / (r1 + innerTan*r2);
+	h.y = (r1*center2->y + innerTan*r2*center1->y) / (r1 + innerTan*r2);
+
+	tan1->x = center1->x + (r1*r1*(h.x-center1->x) + l1*r1*(h.y-center1->y)
+		* sqrt(pow((h.x - center1->x), 2) + pow((h.y - center1->y), 2) - r1*r1)) / pow(distance(h, *center1), 2);
+	tan1->y = center1->y + (r1*r1*(h.y-center1->y) - l1*r1*(h.x-center1->x)
+		* sqrt(pow((h.x-center1->x), 2) + pow((h.y-center1->y), 2) - r1*r1)) / pow(distance(h, *center1), 2);
+
+	tan2->x = center2->x + (r2*r2*(h.x-center2->x) - innerTan*l2*r2*(h.y-center2->y)
+		* sqrt(pow((h.x-center2->x), 2) + pow((h.y-center2->y), 2) - r2*r2)) / pow(distance(h, *center2), 2);
+	tan2->y = center2->y + (r2*r2*(h.y-center2->y) + innerTan*l2*r2*(h.x-center2->x)
+		* sqrt(pow((h.x-center2->x), 2) + pow((h.y-center2->y), 2) - r2*r2)) / pow(distance(h, *center2), 2);
+}
+
+static void computeCenters(point_t *point, float dir, int radius, point_t *center1, point_t *center2) {
+	float dirCos = mcos(dir);
+	float dirSin = msin(dir);
+
+	center1->x = point->x + radius*dirSin;
+	center1->y = point->y - radius*dirCos;
+	center2->x = point->x - radius*dirSin;
+	center2->y = point->y + radius*dirCos;
+}
+
+static struct interpoints computeInterpoints(struct move *depMove, struct move *destMove) {
+	struct interpoints result;
+	int minimum = 0x7FFFFFFF;
+	int i, j, k;
+	float tmp, depDet, destDet;
+	float depDir = depMove->angle*M_PI/128, destDir = destMove->angle*M_PI/128;
+	int startRadius = destMove->startRadius;
+	int endRadius = destMove->endRadius;
+
+	point_t tan1, tan2;
+	point_t depCenters[2], destCenters[2];
+	point_t dep = {depMove->x, depMove->y};
+	point_t dest = {destMove->x, destMove->y};
+
+	// both circles cannot be of the same radius because of the way we compute tangents
+	if(startRadius == endRadius)
+		startRadius++;
+
+	computeCenters(&dep, depDir, startRadius, &depCenters[0], &depCenters[1]);
+	computeCenters(&dest, destDir, endRadius, &destCenters[0], &destCenters[1]);
+
+	for(i=0; i<2; i++) { // try the 2 possible departure circles
```

Comment:
Loop variables should be declared when used, as in
``` c
    for (int i = 0; i < 2; i++)
```

We are in 2017, and this has been introduced in C99.